Unlocked secret units in catalog

zvfrozel opened the pull request:
Secret units don't show up in the catalog even if they're unlocked, this fixes that.

Fixes https://github.com/vEnhance/otis-web/issues/210

vEnhance commented on core/views.py:
```diff
@@ -46,7 +46,22 @@ class UnitGroupListView(LoginRequiredMixin, ListView[Unit]):
+            queryset = Unit.objects.exclude(
+                group__hidden=True, group__bonuslevel__level__gt=student.last_level_seen
+            )
+        elif user.is_staff:
```
The staff check should come before the student check I think. (Most staff are OTIS alum's!)

zvfrozel replied:
This is correct, gone and done that

vEnhance commented on core/views.py:
```diff
@@ -46,7 +46,22 @@ class UnitGroupListView(LoginRequiredMixin, ListView[Unit]):
     def get_queryset(self):
-        queryset = Unit.objects.filter(group__hidden=False)
+        assert isinstance(user := self.request.user, User)
+        student = Student.objects.filter(user=user, semester__active=True).first()
```
I think this has the effect that, e.g. once summer starts, the secret units disappear, and only reappear if the student re-enrolls in OTIS.

I'd suggest that we actually take the `Max` of `last_level_seen` across all the students? (Of course then we should handle the case of empty `Max`.)

vEnhance replied:
Something like the following (please test and make sure it actually works :P )

```python
level : int | None = Student.objects.filter(user=user).aggregate(
    level=Max("last_level_seen")
)["level"]
```

vEnhance reviewed (changes requested):
Looks pretty good, two minor suggestions on logic changes I'd like to see if you agree with, then we can merge this.

zvfrozel commented:
Should be fixed.

> I think this has the effect that, e.g. once summer starts, the secret units disappear, and only reappear if the student re-enrolls in OTIS.

i think the pk logic should handle that unless a new student instance for next semster is created over the summer.
But I changed the code to the Max aggregate logic anyway.

vEnhance commented:
> i think the pk logic should handle that unless a new student instance for next semster is created over the summer.

Hmm, I don't see how. 

> test_hidden fails now!

Fix the code! The test failing is telling you that something is wrong. If we just ignore it by deleting the test, that defeats the purpose of having the test.

In fact this PR should _add_ new tests to check the logic for these secret units.

zvfrozel commented:
> In fact this PR should add new tests to check the logic for these secret units.

Yes, of course it should! I've fixed the broken tests now and added more tests for the new logic

I had left the PR incomplete halfway through yesterday, sorry about that.

vEnhance commented:
> I had left the PR incomplete halfway through yesterday, sorry about that.

ah, okay, mkaes sense. No worries!

This looks great now, thanks for making it happen.
